src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs

```diff
@@ -106,12 +106,36 @@ public ConfidentialClientApplicationBuilder WithClientSecret(string clientSecret
             return this;
         }
 
+        // This is for back-compat with oldstyle API.  Once we deprecate that, we can remove this.
+        internal ConfidentialClientApplicationBuilder WithClientCredential(ClientCredential clientCredential)
+        {
+            Config.ClientCredential = clientCredential;
+            return this;
+        }
+
         /// <inheritdoc />
         internal override ApplicationConfiguration BuildConfiguration()
         {
             base.BuildConfiguration();
 
-            if (!string.IsNullOrWhiteSpace(Config.ClientSecret) && Config.Certificate != null)
+            int numSpecified = 0;
+
+            if (!string.IsNullOrWhiteSpace(Config.ClientSecret))
+            {
+                numSpecified++;
+            }
+
+            if (Config.Certificate != null)
+            {
+                numSpecified++;
+            }
+
+            if (Config.ClientCredential != null)
+            {
+                numSpecified++;
+            }
+
+            if (numSpecified > 1)
             {
                 throw new InvalidOperationException(
                     "ClientSecret and Certificate are mutually exclusive properties.  Only specify one.");
```

Comment:
>ClientSecret and Certificate are mutually exclusive properties.  Only specify one." [](start = 21, length = 83)

aside: we have a string literals file for all error/exception messages

Reply:
I'll add a TODO(migration) here to fix this and ensure we go through and clean all of these up.

---
In reply to: [247263898](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/780#discussion_r247263898) [](ancestors = 247263898)